docs/proposals/resource_management.md

```diff
@@ -0,0 +1,387 @@
+ - If a container exits, all its anonymous pages (tmpfs) gets accounted to the machine (root). With pod level cgroups, that usage can be attributed to the pod.
+ - Let containers share resources - with pod level limits, containers can choose to burst on demand and still be capped at their pod limits.
+
+## Design
+
+High level requirements for the design are as follows:
+ - Do not break existing users. Ideally, there should be no changes to the Kubernetes API semantics.
+ - Support multiple cgroup managers - systemd, cgroupfs, etc.
+
+How we intend to achieve these high level goals is covered in greater detail in the Implementation Plan.
+
+### Proposed cgroup hierarchy:
+
+For the initial implementation we will only support limits for cpu and memory resources.<br>
+
+We propose the following hierarchy:
+
+```
+$ROOT
+  |
+  +- Pod1
+  |   |
+  |   +- Container1
+  |   +- Container2
+  |   ...
+  +- Pod2
+  |   +- Container3
+  |   ...
+  +- ...
+  |
+  +- Bu
+  |   |
+  |   +- Pod3
+  |   |   |
+  |   |   +- Container4
+  |   |   ...
+  |   +- Pod4
+  |   |   +- Container5
+  |   |   ...
+  |   +- ...
+  |
+  +- BE
+  |   |
+  |   +- Pod5
+  |   |   |
+  |   |   +- Container6
+  |   |   +- Container7
+  |   |   ...
+  |   +- ...
+  |
+  +- System-reserved
+  |   |
+  |   +- system
+  |   +- ...
+  |
+  +- Kube-reserved 
+  |   |
+  |   +- kubelet
+  |   +- docker (optional)
+  |   +- ...
+  |
+```
+
+- We have a top level Qos cgroup for the `Bu`(Burstable) and `BE`(Best Effort) Qos classes.
+- But we __dont__ have a separate cgroup for the `G`(Guaranteed) QoS class by default. Guaranteed pod cgroups are by default brought up directly under the `Root` cgroup. Although we have a `guaranteed-qos-at-root` flag to allow users to specify if a separate `G`(Guaranteed) cgroup is required.
+- Each pod has its own cgroup which is nested under the cgroup matching the pod's Qos class.
+- All containers brought up by the pod are nested under the pod's cgroup.
+Examples in the next section.
+- system-reserved cgroup contains the system specific processes.
+- kube-reserved cgroup contains the kubelet specific daemons.
+
+#### Quality of Service classes
+
+A pod can belong to one of the following 3 QoS classes: Guaranteed, Burstable, and Best-Effort, in decreasing order of priority.<br>
+We use the following denotation:<br>
+`G⇒ Guaranteed QoS, Bu⇒ Burstable QoS, BE⇒ Besteffort Qos`
+
+
+#### Guaranteed Cgroup ####
+
+Guaranteed pods will be placed at the `Root` cgroup by default. We would have a flag `guaranteed-qos-at-root` which would allow the user to explicitly specify if he needs a separate `G` cgroup for the Guaranteed Qos class. So by default the flag is set to true.
+
+/PodUID/cpu.quota = cpu limit of Pod
+/PodUID/cpu.shares = cpu request of Pod
+/m.limit_in_bytes = summation of memory limits of all guaranteed containers
```

Comment:
This is incorrect. We don't have to set memory limits for `G` QoS Class. @dubstack can you remove this setting?